fhir-validation/src/main/java/com/ibm/watson/health/fhir/validation/util/ConstraintGenerator.java

```diff
@@ -0,0 +1,182 @@
+
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.StringJoiner;
+import java.util.UUID;
+
+import com.ibm.watson.health.fhir.model.annotation.Constraint;
+import com.ibm.watson.health.fhir.model.resource.StructureDefinition;
+import com.ibm.watson.health.fhir.model.type.Code;
+import com.ibm.watson.health.fhir.model.type.Element;
+import com.ibm.watson.health.fhir.model.type.ElementDefinition;
+import com.ibm.watson.health.fhir.model.type.Uri;
+
+public class ConstraintGenerator {
+    public List<Constraint> generate(StructureDefinition profile) {
+        List<Constraint> constraints = new ArrayList<>();
+        Tree tree = buildTree(profile);
+        for (Node child : tree.root.children) {
+            String expr = transform(child);
+            constraints.add(createConstraint(UUID.randomUUID().toString(), Constraint.LEVEL_RULE, Constraint.LOCATION_BASE, "<no description>", expr, false));
+        }
+        return constraints;
+    }
+    
+    private Tree buildTree(StructureDefinition profile) {
+        Node root = null;
+        
+        Map<String, Node> nodeMap = new LinkedHashMap<>();
+        for (ElementDefinition elementDefinition : profile.getDifferential().getElement()) {
+            String path = elementDefinition.getPath().getValue();
+            
+            Node node = nodeMap.get(path);
+            if (node == null) {
+                node = new Node();
+                node.path = path;
+                
+                int index = path.lastIndexOf(".");
+                node.label = path.substring(index + 1);
+                node.parent = (index != -1) ? nodeMap.get(path.substring(0, index)) : null;
+                
+                if (node.parent == null) {
+                    root = node;
+                } else {
+                    node.parent.children.add(node);
+                }
+                
+                nodeMap.put(path, node);
+            }
+            
+            node.elementDefinitions.add(elementDefinition);
+        }
+        
+        Tree tree = new Tree();
+        tree.root = root;
+        tree.nodeMap = nodeMap;
+        
+        return tree;
+    }
+
+    private String transform(Node node) {
+        StringBuilder sb = new StringBuilder();
+        
+        Integer min = getMin(node);
+        String max = getMax(node);
+        
+        sb.append(node.label);
+        
+        Element fixed = getFixed(node);
+        Element pattern = getPattern(node);
+        if (fixed != null) {
+            sb.append(" = ");
+            if (fixed.is(Code.class)) {
+                sb.append("'").append(fixed.as(Code.class).getValue()).append("'");
+            } else if (fixed.is(Uri.class)) {
+                sb.append("'").append(fixed.as(Uri.class).getValue()).append("'");
+            }
+        } else if (pattern != null) {
+            // TODO: pattern
```

Comment:
is this ok to merge?

Reply:
It’s okay to merge. It’s not hooked up to anything yet.